src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -2201,27 +2202,36 @@ public ArrayType makeArrayType(Type t) {
      * this method could yield surprising answers when invoked on arrays. For example when
      * invoked with t being byte [] and sym being t.sym itself, asSuper would answer null.
      *
-     * @param t a type
-     * @param sym a symbol
-     */
-    public Type asSuper(Type t, Symbol sym) {
-        return asSuper(t, sym, false);
-    }
-
-    /**
-     * Return the (most specific) base type of t that starts with the
-     * given symbol.  If none exists, return null.
+     * Further caveats in Valhalla: There are two "hazards" we need to watch out for when using
+     * this method.
      *
-     * Caveat Emptor: Since javac represents the class of all arrays with a singleton
-     * symbol Symtab.arrayClass, which by being a singleton cannot hold any discriminant,
-     * this method could yield surprising answers when invoked on arrays. For example when
-     * invoked with t being byte [] and sym being t.sym itself, asSuper would answer null.
+     * 1. Since Foo.ref and Foo.val share the same symbol, that of Foo.class, a call to
+     *    asSuper(Foo.ref.type, Foo.val.type.tsym) would return non-null. This is NOT correct
+     *    Foo.val is NOT a super type of Foo.ref either in the language model or in the VM's
+     *    world view. An example of such an hazardous call used to exist in Gen.visitTypeCast.
+     *    When we emit code for  (Foo) Foo.ref.instance a check for whether we really need the
+     *    cast cannot/shouldn't be gated on asSuper(tree.expr.type, tree.clazz.type.tsym) == null)
+     *    but use !types.isSubtype(tree.expr.type, tree.clazz.type) which operates in terms of
+     *    types. When we operate in terms of symbols, there is a loss of type information leading
+     *    to a hazard. Whether a call to asSuper should be transformed into a isSubtype call is
+     *    tricky. isSubtype returns just a boolean while asSuper returns richer information which
+     *    may be required at the call site. Also where the concerned symbol corresponds to a
+     *    generic class, an asSuper call cannot be conveniently rewritten as an isSubtype call
+     *    (see that asSuper(ArrayList<String>.type, List<T>.tsym) != null while
+     *    isSubType(ArrayList<String>.type, List<T>.type) is false;) So care needs to be exercised.
+     *
+     * 2. Given a primitive class Foo, a call to asSuper(Foo.type, SuperclassOfFoo.tsym) and/or
```

Comment:
Question here: is Foo.type refusing to play because the "default" semantics attached to Foo.type is the one of the "value projection" ? If so, then I agree with the fix (e.g. using reference projection or self). Also, given where the model is, I don't think this is an "hack" or a "workaround" - it's basically a requirement: inheritance info lives on the .ref side of the fence.

In fact, so much so that I'll raise: shouldn't this call to "refProjectionOrSelf" be part of asSuper? If asSuper is always use to check for inheritance  relationship, isn't it always the case that you want the reference projection to be used? (and then, if a callsite ends up with too lose a semantics, maybe that's a sign that using asSuper there was the wrong choice?)

Reply:
Plausible, will experiment.